[Counterfactual] Show notification when submitting counterfactual tx

usame-algan opened the pull request:
## What it solves

Resolves #

## How this PR fixes it

- Closes the deploy safe tx-flow when submitting the transaction
- Shows a notification when submitting the deploy safe transaction
- Shows an info screen on the first-tx + deploy flow breaking down the cost of the transaction
- Fixes an issue where the counterfactual state was cleared even if the transaction failed
- Estimates the actual gas costs of the signed transaction and uses that instead of what `useDeployGasLimit` returns

## How to test it

1. Create a counterfactual safe
2. Choose the Activate now option from settings
3. Submit the transaction
4. Observe the modal closes
5. Observe there is a notification
6. Deploy the safe with the first transaction
7. Observe there is an info screen visible
8. Cancel the transaction in your wallet
9. Observe an error notification shown
10. The counterfactual state should not be cleared in that case

## Screenshots

<img width="365" alt="Screenshot 2024-02-08 at 17 54 18" src="https://github.com/safe-global/safe-wallet-web/assets/5880855/a772fa2d-5724-424a-a0cd-2f61b85d3c6b">
<img width="706" alt="Screenshot 2024-02-09 at 12 29 55" src="https://github.com/safe-global/safe-wallet-web/assets/5880855/9d1b0a12-85e9-4293-80db-f37dd0888cd9">


## Checklist
* [ ] I've tested the branch on mobile 📱
* [ ] I've documented how it affects the analytics (if at all) 📊
* [ ] I've written a unit/e2e test for it (if applicable) 🧑‍💻

katspaugh commented on src/features/counterfactual/CounterfactualForm.tsx:
```diff
@@ -113,6 +114,29 @@ export const CounterfactualForm = ({
+            )}{' '}
+            {chain?.nativeCurrency.symbol}
+          </strong>
+          . This is an estimation and the final cost can be slightly higher.
```
```suggestion
          . This is an estimation and the final cost can be higher.
```

katspaugh commented on src/features/counterfactual/CounterfactualForm.tsx:
```diff
@@ -141,15 +164,12 @@ export const CounterfactualForm = ({
         )}
-
         <Divider className={commonCss.nestedDivider} sx={{ pt: 3 }} />
-
```
Waren gute, diese.

usame-algan replied:
This also happened another time but I can't figure out how they are deleted. Afaik we don't have any lint or prettier rules to remove these...

katspaugh commented on src/features/counterfactual/utils.ts:
```diff
@@ -178,3 +182,16 @@ export const createCounterfactualSafe = (
+      variant: 'info',
+      groupKey: 'cf-activate-account',
+      message: 'Transaction submitted',
+      detailedMessage: 'Your Safe Account will be deployed on-chain after the transaction has been executed.',
```
```suggestion
      detailedMessage: 'Your Safe Account will be deployed onchain after the transaction is executed.',
```

schmanu commented on src/features/counterfactual/CounterfactualForm.tsx:
```diff
@@ -113,6 +114,30 @@ export const CounterfactualForm = ({
+            )}{' '}
+            {chain?.nativeCurrency.symbol}
+          </strong>
+          , activation fee is{' '}
```
Should we add here that this fee will only have to be paid once or that future transactions will not have to pay this?

usame-algan replied:
Good idea, I reworded it to say "one time activation fee is ..."

schmanu commented on src/features/counterfactual/utils.ts:
```diff
@@ -65,8 +67,11 @@ export const dispatchTxExecutionAndDeploySafe = async (
     const deploymentTx = await sdkUnchecked.wrapSafeTransactionIntoDeploymentBatch(signedTx, txOptions)
 
+    // We need to estimate the actual gasLimit after the user has signed since it is more accurate than what useDeployGasLimit returns
+    const gas = await signer.estimateGas({ data: deploymentTx.data, value: deploymentTx.value, to: deploymentTx.to })
```
Should we compare this to the estimation and give a prompt / warning if the difference is more than e.g. 50k gas?

usame-algan replied:
I like the idea but this would only happen after the user has pressed Sign in their wallet. I am not sure how helpful it would be to see a warning somewhere in our UI at that point so I would leave it out for now and hope there are no substantial differences between those first transactions. Worst case we have the info text to warn the user about higher cost on execution.

schmanu commented on src/features/counterfactual/hooks/useDeployGasLimit.ts:
```diff
@@ -21,24 +21,33 @@ const getExtraGasForSafety = (safeTx?: SafeTransaction): bigint => {
     ])
 
-    return BigInt(gas) + BigInt(safeTxGas) + BigInt(safeDeploymentGas) + extraGasForSafety
+    const totalGas = BigInt(baseGas) + BigInt(safeTxGas) + BigInt(safeDeploymentGas) + extraGasForSafety
```
Let's add a test for this hook

francovenica commented:
Issue/Question:
1 - In this PR I cannot create a tx. I tried send funds and adding owners, but they always fail. I tried increasing the gas fee, but is to no avail.
Are regular tx working? I always get the same error. I want to verify the kind of notifications I see when I execute a send funds tx but currently I can't.
![image](https://github.com/safe-global/safe-wallet-web/assets/4503659/1763aa51-c25e-4f3c-8781-a6edc529d9ee)

usame-algan commented:
> In this PR I cannot create a tx. I tried send funds and adding owners, but they always fail. I tried increasing the gas fee, but is to no avail.

There was a small issue with our last commit but I just pushed a small fix. It should work again now.

francovenica commented:
I can execute tx like adding/swapping owners.

Some feedback (we can discuss what is possible to change or not)

- When you execute the "Activate safe" tx, you get this notification, which is fine, but you have to click in the details to see the text and also in the list of notifications you cannot see the messege, since there is no "details" link to click, you only see a "transaction submitted" which it can be anything.
(This one is fine, but you got to click on details, a minor feedback)
![image](https://github.com/safe-global/safe-wallet-web/assets/4503659/0e3534c3-b847-45c7-b0db-d9f88a893c3e)

This is the same tx, but saying just "tx submitted" doesn't describe that this was the tx that deployed the safe
![image](https://github.com/safe-global/safe-wallet-web/assets/4503659/738404ba-225d-4315-b078-981c18a98200)

- Once the tx that deploys the safe is finally processed and fully executed, I'd expecte a green card notification saying "success" or something. Right now no notification shows up once the safe is finally deployed

- When I execute a tx like swap owner. I only get only the notificaton of "check in blockexplorer" which is the one deploying the safe, but there is no message implying this for the user. I'd expect here for the notification I showed earlier to show up (The "your safe account will be deployed...." notification)
![image](https://github.com/safe-global/safe-wallet-web/assets/4503659/3368176a-b47f-4337-bb19-fed12bf7aeb0)

usame-algan commented:
> When you execute the "Activate safe" tx, you get this notification, which is fine, but you have to click in the details to see the text and also in the list of notifications you cannot see the messege

I adjusted the wording to be more specific
<img width="362" alt="Screenshot 2024-02-12 at 09 34 45" src="https://github.com/safe-global/safe-wallet-web/assets/5880855/4d272fb4-eb88-4d38-9c35-1c852e237e72">

> Once the tx that deploys the safe is finally processed and fully executed, I'd expecte a green card notification saying "success" or something. Right now no notification shows up once the safe is finally deployed

This didn't work because we usually show the green "Success" notification when we detect the transaction in the history which we can't for these transactions because they don't have a txId. I changed it to show the success event when the transaction is successfully executed and when we clear the counterfactual state.

liliya-soroka commented:
@usame-algan , I don't see success notification after the safe was activated :
<img width="468" alt="image" src="https://github.com/safe-global/safe-wallet-web/assets/338622/b42356ee-7f62-449a-a982-a13b253b8645">

usame-algan commented:
> I don't see success notification after the safe was activated :

Thanks, I added a success notification there as well.

liliya-soroka commented:
Verified
